src/libmeasurement_kit/libevent/dns_impl.hpp

```diff
@@ -53,37 +54,66 @@ class QueryContext : public NonMovable, public NonCopyable {
     }
 };
 
+struct evdns_base_deleter {
+    void operator()(evdns_base* p) {
+        constexpr int fail_requests = 1;
+        evdns_base_free(p, fail_requests);
+    }
+};
+
+struct evaddrinfo_deleter {
+    void operator()(evutil_addrinfo* p) {
+        evutil_freeaddrinfo(p);
+    }
+};
+
+using evdns_base_uptr = std::unique_ptr<evdns_base, evdns_base_deleter>;
+using evaddrinfo_uptr = std::unique_ptr<evutil_addrinfo, evaddrinfo_deleter>;
+
 template <MK_MOCK(evdns_base_new), MK_MOCK(evdns_base_nameserver_ip_add),
         MK_MOCK(evdns_base_free), MK_MOCK(evdns_base_set_option)>
 static inline evdns_base *create_evdns_base(
         Settings settings, Var<Reactor> reactor = Reactor::global()) {
 
-    evdns_base *base;
     event_base *evb = reactor->get_event_base();
+    const int initialize_nameservers = settings.count("dns/nameserver") ? 0 : 1;
+    evdns_base_uptr base(evdns_base_new(evb, initialize_nameservers));
+    if (!base) {
+        throw std::bad_alloc();
+    }
 
-    if (settings.find("dns/nameserver") != settings.end()) {
-        if ((base = evdns_base_new(evb, 0)) == nullptr) {
-            throw std::bad_alloc();
+    if (settings.count("dns/nameserver")) {
+        // libevent can't handle link-local IPv6 nameserver in
+        // evdns_base_nameserver_ip_add, and there is no way to parse alike
+        // addresses in platform-independent way, so that's why getaddrinfo().
+        // fe80::227:22ff:fe45:3a92%wlan0 -- Android
+        // fe80::227:22ff:fe45:3a92%wlp1s0 -- Ubuntu Linux
+        evutil_addrinfo hints = { };
```

Comment:
> What g++ version?

4.9.4 (it's quite old, unfortunately it's the only g++ on VoidLinux).

For completeness, that's what I see:

```
$ g++ --version
g++ (GCC) 4.9.4
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
$ cat prova.cpp
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>

int main() {
    addrinfo hints{};
}
$ g++ -Wall -Wextra -std=c++11 -S prova.cpp
prova.cpp: In function 'int main()':
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_flags' [-Wmissing-field-initializers]
     addrinfo hints{};
                    ^
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_family' [-Wmissing-field-initializers]
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_socktype' [-Wmissing-field-initializers]
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_protocol' [-Wmissing-field-initializers]
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_addrlen' [-Wmissing-field-initializers]
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_addr' [-Wmissing-field-initializers]
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_canonname' [-Wmissing-field-initializers]
prova.cpp:6:20: warning: missing initializer for member 'addrinfo::ai_next' [-Wmissing-field-initializers]
prova.cpp:6:14: warning: unused variable 'hints' [-Wunused-variable]
     addrinfo hints{};
              ^
```
> I wonder if it's the difference [...]

I've checked, but even with `= {}` I obtain the same warnings.